Mathlib/Dynamics/TopologicalEntropy/Semiconj.lean

```diff
@@ -0,0 +1,197 @@
+  (limsup_le_limsup) (Eventually.of_forall fun n ↦ (monotone_div_right_of_nonneg
+    (Nat.cast_nonneg' n) (log_monotone (ENat.toENNReal_mono (coverMincard_image_le h F V n)))))
+
+lemma coverEntropyInfEntourage_image_le (h : Semiconj φ S T) (F : Set X) (V : Set (Y × Y)) :
+    coverEntropyInfEntourage T (φ '' F) V ≤ coverEntropyInfEntourage S F ((Prod.map φ φ) ⁻¹' V) :=
+  (liminf_le_liminf) (Eventually.of_forall fun n ↦ (monotone_div_right_of_nonneg
+    (Nat.cast_nonneg' n) (log_monotone (ENat.toENNReal_mono (coverMincard_image_le h F V n)))))
+
+/-- The entropy of `φ '' F` equals the entropy of `F` if `X` is endowed with the pullback by `φ`
+  of the uniform structure of `Y`.-/
+theorem coverEntropy_image (u : UniformSpace Y) {S : X → X} {T : Y → Y} {φ : X → Y}
+    (h : Semiconj φ S T) (F : Set X) :
+    coverEntropy T (φ '' F) = @coverEntropy X (comap φ u) S F := by
+  apply le_antisymm
+  · refine iSup₂_le fun V V_uni ↦ (coverEntropyEntourage_image_le h F V).trans ?_
+    apply @coverEntropyEntourage_le_coverEntropy X (comap φ u) S F
+    rw [uniformity_comap φ, mem_comap]
+    exact ⟨V, V_uni, Subset.refl _⟩
+  · refine iSup₂_le fun U U_uni ↦ ?_
+    simp only [uniformity_comap φ, mem_comap] at U_uni
+    rcases U_uni with ⟨V, V_uni, V_sub⟩
+    rcases comp_symm_mem_uniformity_sets V_uni with ⟨W, W_uni, W_symm, W_V⟩
+    apply (coverEntropyEntourage_antitone S F ((preimage_mono W_V).trans V_sub)).trans
+    apply (le_coverEntropyEntourage_image h F W_symm).trans
+    exact coverEntropyEntourage_le_coverEntropy T (φ '' F) W_uni
+
+/-- The entropy of `φ '' F` equals the entropy of `F` if `X` is endowed with the pullback by `φ`
+  of the uniform structure of `Y`. This version uses a `liminf`.-/
+theorem coverEntropyInf_image (u : UniformSpace Y) {S : X → X} {T : Y → Y} {φ : X → Y}
+    (h : Semiconj φ S T) (F : Set X) :
+    coverEntropyInf T (φ '' F) = @coverEntropyInf X (comap φ u) S F := by
+  apply le_antisymm
+  · refine iSup₂_le fun V V_uni ↦ (coverEntropyInfEntourage_image_le h F V).trans ?_
+    apply @coverEntropyInfEntourage_le_coverEntropyInf X (comap φ u) S F
+    rw [uniformity_comap φ, mem_comap]
+    exact ⟨V, V_uni, Subset.refl _⟩
+  · refine iSup₂_le fun U U_uni ↦ ?_
+    simp only [uniformity_comap φ, mem_comap] at U_uni
+    rcases U_uni with ⟨V, V_uni, V_sub⟩
+    rcases comp_symm_mem_uniformity_sets V_uni with ⟨W, W_uni, W_symm, W_V⟩
+    apply (coverEntropyInfEntourage_antitone S F ((preimage_mono W_V).trans V_sub)).trans
+    apply (le_coverEntropyInfEntourage_image h F W_symm).trans
+    exact coverEntropyInfEntourage_le_coverEntropyInf T (φ '' F) W_uni
+
+/-- The entropy of `φ '' F` is lower than entropy of `F` if  `φ` is uniformly continuous.-/
+theorem coverEntropy_image_le_of_uniformContinuous [UniformSpace X] [UniformSpace Y] {S : X → X}
+    {T : Y → Y} {φ : X → Y} (h : Semiconj φ S T) (h' : UniformContinuous φ) (F : Set X) :
```

Comment:
Is it enough to assume that \phi is uniformly continuous on the union of the iterates of F (or just on F if it is invariant)? I am imagining a situation where \phi is continuous but not uniformly continuous, and F is compact and invariant (which is a pretty common situation). Then the inequality between entropies should hold, but I don't think the current formulation would get it.

Reply:
Good idea. It's still a bit raw, but I added 4 new lemmas :

- `coverEntropy_restrict` is generalized to any subset of an invariant set: `coverEntropy_restrict_subset`/`coverEntropyInf_restrict_subset `
- I leverage this generalization to get `coverEntropy_image_le_of_uniformContinuousOn_invariant`/`coverEntropyInf_image_le_of_uniformContinuousOn_invariant`

The later statement is the inequality of entropies between `F` and `φ '' F` whenever `φ` is uniformly continuous on an invariant subset `G` which contains `F`.

Taking `G = univ`, we recover `coverEntropy_image_le_of_uniformContinuous`. Taking `F = G`, we recover the case where `F` is invariant. Taking `G` as the union of iterates of `F`, we recover your first statement.

I am keeping `coverEntropy_restrict` for its conceptual importance, and `coverEntropy_image_le_of_uniformContinuous` because it's pretty clean (I think this is the one which would most often be used) and helps in the proof of the more general statement. I don't think it's worth writing the other consequences, because everything should be plug-an-play from here, but I can still add them if needed.